File: VodArchiver/VodArchiver/UserInfo/IUserInfo.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Xml;
using VodArchiver.VideoInfo;

namespace VodArchiver.UserInfo {
	public struct FetchReturnValue { public bool Success; public bool HasMore; public long TotalVideos; public int VideoCountThisFetch; public List<IVideoInfo> Videos; }

	public abstract class IUserInfo : IEquatable<IUserInfo>, IEqualityComparer<IUserInfo>, IComparable<IUserInfo> {
		public abstract ServiceVideoCategoryType Type { get; }
		public abstract string UserIdentifier { get; }

		public abstract bool Persistable { get; set; }
		public abstract bool AutoDownload { get; set; }
		public abstract DateTime LastRefreshedOn { get; set; }

		public abstract XmlNode Serialize( XmlDocument document, XmlNode node );

		public static IUserInfo Deserialize( XmlNode node ) {
			ServiceVideoCategoryType type = (ServiceVideoCategoryType)Enum.Parse( typeof( ServiceVideoCategoryType ), node.Attributes["_type"].Value );
			switch ( type ) {
				case ServiceVideoCategoryType.TwitchRecordings:
				case ServiceVideoCategoryType.TwitchHighlights:
					return new TwitchUserInfo( node );
				case ServiceVideoCategoryType.HitboxRecordings:
					return new HitboxUserInfo( node );
				case ServiceVideoCategoryType.YoutubeUser:
					return new YoutubeUserUserInfo( node );
				case ServiceVideoCategoryType.YoutubeChannel:
					return new YoutubeChannelUserInfo( node );
				case ServiceVideoCategoryType.YoutubePlaylist:
					return new YoutubePlaylistUserInfo( node );
				case ServiceVideoCategoryType.RssFeed:
					return new RssFeedUserInfo( node );
				case ServiceVideoCategoryType.FFMpegJob:
					return new FFMpegJobUserInfo( node );
				case ServiceVideoCategoryType.ArchiveOrg:
					return new ArchiveOrgUserInfo(node);
				default:
					throw new Exception( "Unknown user info type: " + type );
			}
		}

		public abstract Task<FetchReturnValue> Fetch( int offset, bool flat );

		public virtual int CompareTo( IUserInfo other ) {
			if ( this.Type != other.Type ) {
				return this.Type.CompareTo( other.Type );
			}

			return this.UserIdentifier.CompareTo( other.UserIdentifier );
		}

		public virtual bool Equals( IUserInfo other ) {
			return this.Type == other.Type && this.UserIdentifier == other.UserIdentifier;
		}

		public virtual bool Equals( IUserInfo x, IUserInfo y ) {
			return x.Equals( y );
		}

		public override int GetHashCode() {
			return this.UserIdentifier.GetHashCode() ^ this.Type.GetHashCode();
		}

		public virtual int GetHashCode( IUserInfo obj ) {
			return obj.GetHashCode();
		}

		public override string ToString() {
			return Type + ": " + UserIdentifier;
		}
	}
}

File: native/vodarchiver/userinfo/i-user-info.cpp
#include "i-user-info.h"

#include <format>
#include <optional>
#include <string_view>

namespace VodArchiver {
std::string_view ServiceVideoCategoryTypeToString(ServiceVideoCategoryType type) {
    switch (type) {
        case ServiceVideoCategoryType::TwitchRecordings: return "TwitchRecordings";
        case ServiceVideoCategoryType::TwitchHighlights: return "TwitchHighlights";
        case ServiceVideoCategoryType::HitboxRecordings: return "HitboxRecordings";
        case ServiceVideoCategoryType::YoutubeUser: return "YoutubeUser";
        case ServiceVideoCategoryType::YoutubeChannel: return "YoutubeChannel";
        case ServiceVideoCategoryType::YoutubePlaylist: return "YoutubePlaylist";
        case ServiceVideoCategoryType::RssFeed: return "RssFeed";
        case ServiceVideoCategoryType::FFMpegJob: return "FFMpegJob";
        case ServiceVideoCategoryType::ArchiveOrg: return "ArchiveOrg";
        case ServiceVideoCategoryType::YoutubeUrl: return "YoutubeUrl";
        default: return "Unknown";
    }
}

std::optional<ServiceVideoCategoryType> ServiceVideoCategoryTypeFromString(std::string_view sv) {
    if (sv == "TwitchRecordings") {
        return ServiceVideoCategoryType::TwitchRecordings;
    } else if (sv == "TwitchHighlights") {
        return ServiceVideoCategoryType::TwitchHighlights;
    } else if (sv == "HitboxRecordings") {
        return ServiceVideoCategoryType::HitboxRecordings;
    } else if (sv == "YoutubeUser") {
        return ServiceVideoCategoryType::YoutubeUser;
    } else if (sv == "YoutubeChannel") {
        return ServiceVideoCategoryType::YoutubeChannel;
    } else if (sv == "YoutubePlaylist") {
        return ServiceVideoCategoryType::YoutubePlaylist;
    } else if (sv == "RssFeed") {
        return ServiceVideoCategoryType::RssFeed;
    } else if (sv == "FFMpegJob") {
        return ServiceVideoCategoryType::FFMpegJob;
    } else if (sv == "ArchiveOrg") {
        return ServiceVideoCategoryType::ArchiveOrg;
    } else if (sv == "YoutubeUrl") {
        return ServiceVideoCategoryType::YoutubeUrl;
    }
    return std::nullopt;
}

IUserInfo::~IUserInfo() = default;

std::string IUserInfo::ToString() {
    return std::format("{}: {}", ServiceVideoCategoryTypeToString(GetType()), GetUserIdentifier());
}
} // namespace VodArchiver
